src/mame/sun/sun3.cpp

```diff
@@ -809,6 +814,23 @@ void sun3_state::vmetype3space_map(address_map &map)
 {
 }
 
+uint32_t sun3_state::enable_r()
+{
+	// Incorporate diag switch value.
+	const uint32_t diagsw = (m_diagsw->read() != 0) ? 0x01000000 : 0;
+	uint32_t data = m_enable | (diagsw);
+	return data;
+}
+
+void sun3_state::enable_w(uint32_t data, uint32_t mem_mask)
+{
+	// Mask out diag switch value, it comes from the switch on read.
+	const uint32_t diag = 0x01000000;
+	const uint32_t diag_norm_mask = ~diag;
+	COMBINE_DATA(&m_enable);
+	m_enable &= diag_norm_mask;
+}
```

Comment:
You can still end up with an issue doing it this way if someone injects a value with the bit set via a debugger memory viewer or something.  You’re better off masking the value on read – something like:

```c++
uint32_t sun3_state::enable_r()
{
	// Incorporate diag switch value.
	const uint32_t diagsw = m_diagsw->read() << 24;
	return m_enable & ~(u32(1) << 24) | diagsw;
}
```

and don’t bother with `enable_w` at all.